keps/sig-architecture/3136-beta-apis-off-by-default/README.md

```diff
@@ -0,0 +1,286 @@
+if v1beta.some.group is currently enabled by default and we create v1beta2.some.group, v1beta2.some.group will still be enabled by default.
+
+## Motivation
+
+Beta APIs are not considered stable and reliance upon APIs in this state leads to exposure to bugs,
+guaranteed migration pain for users when the APIs move to stable, and the risk that dependencies will
+grow around unfinished APIs.
+Enabling beta APIs by default, exacerbates these problems by making them on in nearly every cluster.
+We observed these problems as we removed long-standing beta APIs and the PRR survey tells us that over
+90% of production clusters leave these APIs enabled.
+Unsuitability for production use is documented at https://kubernetes.io/docs/reference/using-api/#api-versioning 
+("The software is not recommended for production uses"), but defaulting on means they are present in nearly every 
+production cluster.
+By disabling beta APIs by default, a cluster-admin can opt-in for specific APIs without having every
+incomplete API present in the cluster.
+
+### Goals
+
+1. Disable new beta APIs by default.
+2. Continue enabling existing beta APIs and new version of existing beta APIs by default:
+   if v1beta.some.group is currently enabled by default and we create v1beta2.some.group, v1beta2.some.group will still be enabled by default.
+3. Allow enabling specific resources in beta.  Enable coolnewjobtype.v1beta1.batch.k8s.io without enabling other-neat-job.v1beta1.batch.k8s.io
+
+### Non-Goals
+
+1. Change featuregate defaults.
+
+## Proposal
+
+New beta APIs will be placed into the `DisableVersions` stanza instead of the `EnableVersions` stanza (see [DefaultAPIResourceConfigSource](https://github.com/kubernetes/kubernetes/blob/0669da445fa8c1ae07c15c0827f0e83da11cbe58/pkg/controlplane/instance.go#L643)).
+The `--runtime-config` flag will be extended to allow `group/version/resource=true`, to enable specific resources.
+To enable a beta API, a cluster-admin will have to add the appropriate `--runtime-config` flags.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a cluster-admin I want to enable the coolnewjobtype.v1beta1.batch.k8s.io API in my cluster.
+
+To do this I call `kube-apiserver --runtime-config=batch.k8s.io/v1beta1/coolnewjobtype`.
+
+#### Story 2
+
+As a cluster-admin I want to enable all beta APIs as in past releases.
+
+To do this I call `kube-apiserver --runtime-config=api/beta=true`.
+This already exists and will continue to function.
+
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+Adoption of beta features will slow.
+Given how kubernetes is now treated, this is a good thing, not a bad thing.
+Those users that want to move quickly and get new features can do so by enabling all beta feature
+or just enabling those that are important for their workload.
+The [PRR survey](https://datastudio.google.com/reporting/2e9c7439-202b-48a9-8c57-4459e0d69c8d/page/Cv5HB) shows that 
```

Comment:
this link request access , 

> over 30% of production clusters have alpha features enabled

that is not a justification per se without explaining how the request access, I have experience from other projects using these kind of surveys to justify decisions, and most of the data was faked or invalid

Reply:
> this link request access ,

Access added.

The data doesn't appear to support your hypothesis given that cluster-admins in every range from 0-10, 10-100, 101-1000, 1001-10000, more than 10k all have about a 30% rate of enabling alpha features.  I think the phrasing  here as a conclusion  is appropriate.

Reply:
> The data doesn't appear to support your hypothesis given that cluster-admins in every range from 0-10, 10-100, 101-1000, 1001-10000, more than 10k all have about a 30% rate of enabling alpha features. I think the phrasing here as a conclusion is appropriate.

I was trying to provide an example on why a survey alone is not valid to obtain conclusions and I failed because as you demonstrate my example was just against the survey.

> over 30% of production clusters have alpha features enabled, so cluster-admins are willing and able to enable features

how many clusters were surveyed ? how many cluster do we estimate can be running in the world?
is this a 1% of the total population of clusters or a 90%? ...

that is a correlation based on the survey data, but correlation doesn't mean causality :smile: 

I think that the correct wording is "based on the data from the survey we have some indicators that cluster-admins are willing and able to enable features that are not on by default when they are desired." 

But at this point I feel I'm bike-shedding, is just that I don't like using surveys to justify decisions without strong correlation, feel free to discard the comment

Reply:
I can't believe 30% of clusters have some alpha features enabled, how many clusters total are represented in the survey? I suspect substantial sampling bias.

Reply:
Also note that "30% of clusters have at least one alpha feature enabled" even if true, is NOT equivalent to the thing we really care about, which is "a new pre-GA feature will (with high probability) get exercised prior to GA". (Since the clusters could all be using the same alpha feature.)

Reply:
(That said, the actual point in the text is just that it demonstrably actually happens, so the text is not wrong. That's just a very low bar.)

Reply:
The core claim is "cluster-admins are willing and able to enable features".

It doesn't sound like there is a dispute on that point. I've updated the text to try to make the claim more clear.

Reply:
AWS EKS doesn't enable cluster admins to change feature gates at the moment (https://github.com/aws/containers-roadmap/issues/512).

Reply:
It's not "30% of clusters have alpha features enabled". It's "30% of cluster admins have enabled an alpha feature in production". Very different claims. But still supporting what we need.